private/pkg/storage/storagegit/bucket.go

```diff
@@ -0,0 +1,185 @@
+	objectReader git.ObjectReader,
+	root git.Tree,
+) (storage.ReadBucket, error) {
+	return &bucket{
+		objectReader: objectReader,
+		root:         root,
+	}, nil
+}
+
+func (b *bucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	entry, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch entry.Mode() {
+	case git.ModeFile, git.ModeExe:
+		data, err := b.objectReader.Blob(entry.Hash())
+		if err != nil {
+			return nil, err
+		}
+		return &namedReader{
+			info:   b.newObjectInfo(path),
+			reader: bytes.NewReader(data),
+		}, nil
+	case git.ModeSymlink:
+		// Symlinks are stored as blobs that reference the target path as a relative
+		// path. We can follow this symlink trivially.
+		data, err := b.objectReader.Blob(entry.Hash())
+		if err != nil {
+			return nil, err
+		}
+		path = normalpath.Join(
+			normalpath.Base(path),
+			normalpath.Normalize(string(data)),
+		)
+		return b.Get(ctx, path)
+	default:
+		return nil, storage.NewErrNotExist(path)
+	}
+}
+
+func (b *bucket) Stat(ctx context.Context, path string) (storage.ObjectInfo, error) {
+	entry, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch entry.Mode() {
+	case git.ModeFile, git.ModeExe, git.ModeSymlink:
+		return b.newObjectInfo(path), nil
+	default:
+		// TODO: should this be an error? What kind of error can we throw here?
+		return nil, storage.NewErrNotExist(path)
+	}
+}
+
+func (b *bucket) Walk(ctx context.Context, prefix string, f func(storage.ObjectInfo) error) error {
+	walkChecker := storageutil.NewWalkChecker()
+	return b.walk(b.root, b.objectReader, prefix, func(path string, te git.TreeEntry) error {
+		if err := walkChecker.Check(ctx); err != nil {
+			return err
+		}
+		return f(b.newObjectInfo(path))
+	})
+}
+
+func (b *bucket) walk(
+	root git.Tree,
+	objectReader git.ObjectReader,
+	prefix string,
+	walkFn func(string, git.TreeEntry) error,
+) error {
+	prefix = normalpath.Normalize(prefix)
+	if prefix != "." {
+		entry, err := b.root.Traverse(b.objectReader, normalpath.Components(prefix)...)
+		if err != nil {
+			if errors.Is(err, git.ErrSubTreeNotFound) {
+				return storage.NewErrNotExist(prefix)
+			}
+			return err
+		}
+		subTree, err := b.objectReader.Tree(entry.Hash())
+		if err != nil {
+			return err
+		}
+		root = subTree
+	}
+	return walkTree(root, objectReader, prefix, walkFn)
+}
+
+func walkTree(
+	root git.Tree,
+	objectReader git.ObjectReader,
+	prefix string,
+	walkFn func(string, git.TreeEntry) error,
+) error {
+	for _, entry := range root.Entries() {
+		path := normalpath.Join(prefix, entry.Name())
+		switch entry.Mode() {
+		case git.ModeFile, git.ModeExe, git.ModeSymlink:
```

Comment:
Probably an option in the bucket constructor that defines if we should follow symlinks or not?